pkg/cloudprovider/aws/capacity.go

```diff
@@ -35,34 +49,57 @@ type Capacity struct {
 	instanceTypeProvider   *InstanceTypeProvider
 }
 
+// AWSConstraints are AWS specific constraints
+type AWSConstraints struct {
+	cloudprovider.Constraints
+	CapacityType string
+}
+
+func NewAWSConstraints(constraints *cloudprovider.Constraints) *AWSConstraints {
+	capacityType, ok := constraints.Labels[capacityTypeLabel]
+	if !ok {
+		capacityType = capacityTypeOnDemand
+	}
+	return &AWSConstraints{
+		Constraints:  *constraints,
+		CapacityType: capacityType,
+	}
+}
+
 // Create a set of nodes given the constraints.
 func (c *Capacity) Create(ctx context.Context, constraints *cloudprovider.Constraints) ([]cloudprovider.Packing, error) {
-	// 1. Retrieve normalized zones from constraints or all zones that the cluster spans if no zonal constraints are specified
-	zonalSubnetOptions, err := c.vpcProvider.GetZonalSubnets(ctx, constraints, c.spec.Cluster.Name)
+	// 1. Create AWS Cloud Provider constraints and apply defaults
+	awsConstraints := NewAWSConstraints(constraints)
+
+	// 2. Retrieve normalized zones from constraints or all zones that the cluster spans if no zonal constraints are specified
+	zonalSubnetOptions, err := c.vpcProvider.GetZonalSubnets(ctx, awsConstraints, c.spec.Cluster.Name)
 	if err != nil {
 		return nil, fmt.Errorf("getting zonal subnets, %w", err)
 	}
 
-	// 2. Filter for instance types that fit constraints
-	zonalInstanceTypes, err := c.instanceTypeProvider.Get(ctx, zonalSubnetOptions, constraints)
+	// 3. Filter for instance types that fit constraints
+	zonalInstanceTypes, err := c.instanceTypeProvider.Get(ctx, zonalSubnetOptions, awsConstraints)
 	if err != nil {
 		return nil, fmt.Errorf("filtering instance types by constraints, %w", err)
 	}
 
-	// 3. Compute Packing given the pods and instance types
-	instancePackings := c.packer.Pack(ctx, constraints.Pods, zonalInstanceTypes, constraints)
+	// 4. Compute Packing given the pods and instance types
+	instancePackings := c.packer.Pack(ctx, awsConstraints.Pods, zonalInstanceTypes, constraints)
 
-	zap.S().Debugf("Computed packings for %d pod(s) onto %d node(s)", len(constraints.Pods), len(instancePackings))
-	launchTemplate, err := c.launchTemplateProvider.Get(ctx, c.spec.Cluster, constraints)
+	zap.S().Debugf("Computed packings for %d pod(s) onto %d node(s)", len(awsConstraints.Pods), len(instancePackings))
+	launchTemplate, err := c.launchTemplateProvider.Get(ctx, c.spec.Cluster, awsConstraints)
 	if err != nil {
 		return nil, fmt.Errorf("getting launch template, %w", err)
 	}
 
-	// 4. Create Instances
+	// 5. Create Instances
 	var instanceIDs []*string
 	podsForInstance := make(map[string][]*v1.Pod)
 	for _, packing := range instancePackings {
-		instanceID, err := c.instanceProvider.Create(ctx, launchTemplate, packing.InstanceTypes, zonalSubnetOptions)
+		if len(packing.InstanceTypes) > maxInstanceTypes {
+			packing.InstanceTypes = packing.InstanceTypes[:maxInstanceTypes]
```

Comment:
Could use a comment explaining why this is needed? Also is it guaranteed that the packings "at the end" are the best ones to discard?

Reply:
Sure, I'll add a comment explaining why this is needed, but it's basically because the fleet request size can grow to be too large and get rejected by the EC2 API, if we pass in say 200+ instance types.  I started it at 10 (which will pass in `10 instance types * azs to fleet) since that is still quite a flexible request.  I'm not opposed to changing the number though to something else. 

And yes, it is guaranteed that the instanceTypes packings at the end are the best ones to drop from a point-in-time packing efficiency perspective. The packing pkg returns the instance types sorted by ascending resources.

Reply:
Ascending resources is a bit tricky since it's high dimensional. Do we want to prioritize CPU? Memory? Euclidean?

Reply:
it only sorts by CPU (primary) and memory (secondary). I'll make the explicit in the comments rather than keeping it generic for now until we add another sort parameter.

Reply:
I think it's worth some discussion. I think we probably prefer to have tight memory packing and extra CPU, since CPU is more burstable than memory. You could also imagine cases where we're selecting much too large instances if there are multiple ICE among one dimension, so a euclidean could help there.

Reply:
I implemented a euclidean distance sort on vcpus and memory and it does a fairly good job of mirroring a price sort without the extra complexity of calling the pricing apis.  I'm not sure I understand your point on getting "ICE[d] among one dimension".  I don't think ICEs would be correlated like that.  

I'm also not sure we should be too picky on what we're sorting on since w/ OD, we don't do any sorting. The spot priorities are best effort and still primarily use capacity (i.e. pool depth) when deciding.

Reply:
Thinking a touch further, if we had pricing data (oof: historical pricing), then we should just price sort. 

Maybe we could implement it generically as a price estimate sort, and then cloud providers can use any mechanism to generate those prices (i.e. calling price history, euclidean distance, etc)

Reply:
That might make sense, it's not really price it's just an arbitrary prioritization that the cloud provider decides.  I don't know if it's worth adding a plug into the cloud provider though... the cloud provider could just resort based on whatever they want when they get the list back from packing.